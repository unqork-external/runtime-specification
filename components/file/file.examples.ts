import { createExamples } from '../../../utilities'

export const filePatternExamples = createExamples({
  standardExamples: ['*', '.txt, .csv, .xlsx'],
})
