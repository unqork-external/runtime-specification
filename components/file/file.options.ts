import { Default, Description } from '@tsed/schema'

import { filePatternExamples } from './file.examples'
import { Examples } from '../../../decorators/schema/examples.decorator'

/**
 * R1 defaults
 */
const FILE_DEFAULTS = {
  browseLink: 'browse',
  compressFile: true,
  dragDropNotSupported: 'File Drag/Drop is not supported for this browser',
  dropFilesLabel: 'Drag & drop your files here, or',
  fileNameLabel: 'File',
  parseOnUpload: false,
  removeLink: 'remove file',
  sizeLabel: 'Size',
  startingUploadMessage: 'Starting upload',
  uqMaxFileSizeMB: '',
  filePattern: '*',
  multiTabExcel: false,
}

@Description('Settings specific to the File component')
export class FileOptions {
  @Description('Whether a user can select multiple files to upload')
  multiple?: boolean

  @Description('Whether uploaded image files should render a preview')
  showImagePreview?: boolean

  @Description('The maximum width in pixels of image previews to show if previews are enabled')
  imagePreviewSize?: number

  @Default(FILE_DEFAULTS.startingUploadMessage)
  @Description('Text shown to users when upload begins')
  startingUploadMessage: string = FILE_DEFAULTS.startingUploadMessage

  @Default(FILE_DEFAULTS.dropFilesLabel)
  @Description('Text label inside the boundaries of the drag and drop area')
  dropFilesLabel: string = FILE_DEFAULTS.dropFilesLabel

  @Default(FILE_DEFAULTS.browseLink)
  @Description('Link text to browse computer file system; appended to end of dropFilesLabel')
  browseLink: string = FILE_DEFAULTS.browseLink

  @Default(FILE_DEFAULTS.dragDropNotSupported)
  @Description("Text to display if user's browser doesn't support drag and drop")
  dragDropNotSupported: string = FILE_DEFAULTS.dragDropNotSupported

  @Default(FILE_DEFAULTS.fileNameLabel)
  @Description('The header label for the column listing the individual file names')
  fileNameLabel: string = FILE_DEFAULTS.fileNameLabel

  @Default(FILE_DEFAULTS.removeLink)
  @Description('Aria-label passed to remove button inside uploaded files table')
  removeLink: string = FILE_DEFAULTS.removeLink

  @Default(FILE_DEFAULTS.sizeLabel)
  @Description('The header label for the column listing the file sizes')
  sizeLabel: string = FILE_DEFAULTS.sizeLabel

  /**
   * Validation Settings
   */
  @Default(FILE_DEFAULTS.uqMaxFileSizeMB)
  @Description('Maximum file size in megabytes')
  uqMaxFileSizeMB: number | string = FILE_DEFAULTS.uqMaxFileSizeMB

  @Description('A quick tip explaining the file size limits of your File component')
  maxFileSizeMessage?: string

  @Description('Error message below uploaded files that exceed the maximum size')
  maxFileSizeWarning?: string

  @Description(`User provided valid file types as a list of comma delimited strings. Use '*' to accept all file types.`)
  @Default(FILE_DEFAULTS.filePattern)
  @Examples(filePatternExamples)
  filePattern: string = FILE_DEFAULTS.filePattern

  @Description('Maximum accepted number of files')
  maxNumberOfFiles?: number

  /**
   * Advanced Settings
   */
  @Default(FILE_DEFAULTS.compressFile)
  @Description('Compresses images if enabled (only supports .jpg and .png image formats)')
  compressFile: boolean = FILE_DEFAULTS.compressFile

  @Default(FILE_DEFAULTS.multiTabExcel)
  @Description('If user uploads a multi-tab Excel file, parse all tabs into one JSON structure')
  multiTabExcel: boolean = FILE_DEFAULTS.multiTabExcel

  @Default(FILE_DEFAULTS.parseOnUpload)
  @Description('File data encoded as Base64 string as part of submission data')
  parseOnUpload: boolean = FILE_DEFAULTS.parseOnUpload

  @Description('When parseOnUpload is TRUE, this character will be used to separate values.')
  delimiterChar?: string
}
