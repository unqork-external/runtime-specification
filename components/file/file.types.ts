export type UploadFile = Pick<File, 'name' | 'size' | 'type'> & {
  data: Blob
  id: string
  url: string
  storage: 'base64' | 'cloud'
  // raw is needed for cloud upload
  raw?: File
  // in case we parse e.g csv files it will generate this
  json?: string
  headings?: string
}

export type FileSubmissionValue = {
  storage: string
  name: string
  size: number
  type: string
  id?: string
  url?: string
  data?: Blob
  // in case we parse e.g csv files it will generate this
  json?: string
  headings?: string
}
