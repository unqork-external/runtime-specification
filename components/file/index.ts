export { FileComponentDefinition } from './file.definition'
export { FileOptions } from './file.options'
export { FileStyleTargets } from './styling/fileStyle.targets'
export { FileTargets, FileStyleClasses } from './targets/fileTargets.enum'
export type { UploadFile, FileSubmissionValue } from './file.types'
