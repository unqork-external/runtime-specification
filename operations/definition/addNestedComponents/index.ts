export * from './addNestedComponents.model'
export * from './addNestedComponents.options'
