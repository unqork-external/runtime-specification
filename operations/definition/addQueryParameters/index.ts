export * from './addQueryParameters.model'
export * from './addQueryParameters.options'
