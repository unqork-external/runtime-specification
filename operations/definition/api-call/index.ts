export * from './apiCall.type'
export * from './method.type'
export * from './serviceType.type'
export * from './apiCall.model'
export * from './apiCall.options'
