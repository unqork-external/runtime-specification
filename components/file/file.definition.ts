import { Const, Default, Description, DiscriminatorValue, Optional, Property, Required } from '@tsed/schema'

import { BaseComponentDefinition } from '../../baseComponentInterface/base.component.definition'
import { Display } from '../../componentComposition/display/component.display'
import { InputField } from '../../componentComposition/input/component.input'
import { Validation } from '../../validations/validation'

export type UploadFile = Pick<File, 'name' | 'size' | 'type'> & {
  data: Blob
  id: string
  url: string
  storage: 'base64' | 'cloud'
}

export type FileSubmissionValue = {
  storage: string
  name: string
  size: number
  type: string
  id?: string
  url?: string
  data?: Blob
}

/**
 * R1 defaults
 */
const DEFAULT_VALUES = {
  browseLink: 'browse',
  compressFile: true,
  dragDropNotSupported: 'File Drag/Drop is not supported for this browser',
  dropFilesLabel: 'Drag & drop your files here, or',
  fileNameLabel: 'File',
  parseOnUpload: false,
  removeLink: 'remove file',
  sizeLabel: 'Size',
  startingUploadMessage: 'Starting upload',
  uqMaxFileSizeMB: '',
}

export class FileOptions {
  /**
   * Display Settings
   */
  @Optional()
  @Description('Whether a user can select multiple files to upload')
  multiple?: boolean

  @Optional()
  @Description('Whether uploaded image files should render a preview')
  showImagePreview?: boolean

  @Description('Whether uploaded image files should render a preview')
  imagePreviewSize: number

  @Default(DEFAULT_VALUES.startingUploadMessage)
  @Description('Descriptive text shown to users when upload begins')
  startingUploadMessage: string = DEFAULT_VALUES.startingUploadMessage

  @Default(DEFAULT_VALUES.dropFilesLabel)
  @Description('Text label inside the boundaries of the drag and drop area')
  dropFilesLabel: string = DEFAULT_VALUES.dropFilesLabel

  @Default(DEFAULT_VALUES.browseLink)
  @Description('Link text to browse computer file system; appended to end of dropFilesLabel')
  browseLink: string = DEFAULT_VALUES.browseLink

  @Default(DEFAULT_VALUES.dragDropNotSupported)
  @Description("Display text if user's browser doesn't support drag and drop")
  dragDropNotSupported: string = DEFAULT_VALUES.dragDropNotSupported

  @Default(DEFAULT_VALUES.fileNameLabel)
  @Description('The header label for the column listing the individual file names')
  fileNameLabel: string = DEFAULT_VALUES.fileNameLabel

  @Default(DEFAULT_VALUES.removeLink)
  @Description('Aria-label passed to remove button inside uploaded files table')
  removeLink: string = DEFAULT_VALUES.removeLink

  @Default(DEFAULT_VALUES.sizeLabel)
  @Description('The header label for the column listing the file sizes')
  sizeLabel: string = DEFAULT_VALUES.sizeLabel

  /**
   * Action Settings
   */
  @Optional()
  @Description('The component to trigger post processing. Value is a Property ID.')
  postTrigger?: string

  /**
   * Validation Settings
   */
  @Default(DEFAULT_VALUES.uqMaxFileSizeMB)
  @Description('Maximum file size in megabytes')
  uqMaxFileSizeMB: number | string = DEFAULT_VALUES.uqMaxFileSizeMB

  @Optional()
  @Description('A quick tip explaining the file size limits of your File component')
  maxFileSizeMessage?: string

  @Optional()
  @Description('Error message below uploaded files that exceed the maximum size')
  maxFileSizeWarning?: string

  @Description('User provided valid file types (comma delimited string list)')
  @Default('*')
  filePattern: string = '*'

  @Optional()
  @Description('Maximum accepted number of files')
  maxNumberOfFiles?: number

  /**
   * Advanced Settings
   */

  @Default(DEFAULT_VALUES.compressFile)
  @Description('Images compression (only supports .jpg and .png image formats)')
  compressFile: boolean = DEFAULT_VALUES.compressFile

  @Description('If user uploads a multi-tab Excel file, parse all tabs into one JSON structure')
  multiTabExcel: boolean

  @Default(DEFAULT_VALUES.parseOnUpload)
  @Description('File data encoded as Base64 string as part of submission data')
  parseOnUpload: boolean = DEFAULT_VALUES.parseOnUpload

  @Optional()
  @Description('Custom CSS class applied to element')
  customClass?: string
}

@DiscriminatorValue('file')
export class FileDefinition extends BaseComponentDefinition {
  @Required()
  @Const('file')
  @Description('Describes the File type')
  type: string = 'file' as const

  @Property(Display)
  display: Display = new Display()

  @Property(InputField)
  @Description('Field settings for the File component.')
  field: InputField = new InputField()

  @Description('Settings specific to the File component')
  options: FileOptions = new FileOptions()

  @Property(Validation)
  @Description('Validation rules')
  validation: Validation = new Validation()

  // No decorator, this is local state for the File component
  fileData?: UploadFile[]

  @Description('Value of the File component. Will be undefined if not set.')
  declare value?: Array<FileSubmissionValue>
}
