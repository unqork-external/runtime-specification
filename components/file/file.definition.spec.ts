import { FileComponentDefinition } from './file.definition'
import { generateSchemaAndValidate } from '../../../utilities'

describe('File Contract Tests', function () {
  it('should ensure the contract is clear', function () {
    const { schema } = generateSchemaAndValidate(FileComponentDefinition)

    expect(schema.required).toMatchSnapshot()
    expect(schema.properties).toMatchSnapshot()
    expect(schema.definitions).toMatchSnapshot()
  })
})
