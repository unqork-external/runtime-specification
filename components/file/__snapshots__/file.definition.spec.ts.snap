// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Contract Tests should ensure the contract is clear 1`] = `
[
  "key",
  "type",
]
`;

exports[`File Contract Tests should ensure the contract is clear 2`] = `
{
  "display": {
    "$ref": "#/definitions/Display",
  },
  "executable": {
    "default": true,
    "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
    "type": "boolean",
  },
  "field": {
    "$ref": "#/definitions/InputField",
  },
  "inputs": {
    "items": {
      "$ref": "#/definitions/InputRef",
    },
    "type": "array",
  },
  "key": {
    "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
    "examples": [
      "I am the key",
      "bestKeyEver",
    ],
    "minLength": 1,
    "type": "string",
  },
  "options": {
    "$ref": "#/definitions/FileOptions",
  },
  "package": {
    "$ref": "#/definitions/PackageModel",
  },
  "signals": {
    "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
    "examples": [
      {
        "args": {
          "inputs": {
            "propertyPath": "value",
            "required": true,
            "targetKey": "firstName",
          },
        },
        "id": "firstNameWatcher",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onWatch",
      },
      {
        "id": "textfieldOnBlur",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "textfield received blur event from DOM",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onBlur",
      },
      "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
    ],
    "type": "object",
  },
  "styling": {
    "$ref": "#/definitions/FileStyling",
  },
  "type": {
    "const": "file",
    "description": "Name of the type of component.",
    "examples": [
      "file",
    ],
    "minLength": 1,
    "type": "string",
  },
  "validation": {
    "$ref": "#/definitions/Validation",
  },
  "value": {
    "description": "Value of the File component. Will be undefined if not set.",
    "type": "array",
  },
}
`;

exports[`File Contract Tests should ensure the contract is clear 3`] = `
{
  "Display": {
    "description": "Basic UI settings related to how a component may be displayed and interacted with. 'Hidden' corresponds to visibility. 'Interactive' corresponds to the end-user's ability to interact with input components. 'Read Only View' corresponds to the component being rendered with an input element or as text.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "Setting it to \`true\` hides the component from the view, while \`false\` shows it. Regardless of this setting, the component is always visible in the Module Editor. A hidden component can still receive information from other components.",
        "type": "boolean",
      },
      "interactive": {
        "default": true,
        "description": "When \`false\`, it is not possible to interact with or edit the input field. Its appearance is modified (grayed-out by default) to show its state, the HTML 'disabled' attribute is added to the field, and when hovering over it, the cursor also changes to alert of the disabled status. However, the component can still receive information from other components when it's not interactive.",
        "type": "boolean",
      },
      "readOnlyView": {
        "default": false,
        "description": "Defines how the component will be displayed. When \`false\`, the component is displayed as an input field. When \`true\`, the component will not accept any input from the user, and its value will be rendered inline as text and not as an input component. If no value is present, the word 'None' is rendered. If a prefix or suffix is set for the component, they also appear as inline text.",
        "type": "boolean",
      },
    },
    "type": "object",
  },
  "FileOptions": {
    "description": "Settings specific to the File component",
    "properties": {
      "browseLink": {
        "default": "browse",
        "description": "Link text to browse computer file system; appended to end of dropFilesLabel",
        "type": "string",
      },
      "compressFile": {
        "default": true,
        "description": "Compresses images if enabled (only supports .jpg and .png image formats)",
        "type": "boolean",
      },
      "dragDropNotSupported": {
        "default": "File Drag/Drop is not supported for this browser",
        "description": "Text to display if user's browser doesn't support drag and drop",
        "type": "string",
      },
      "dropFilesLabel": {
        "default": "Drag & drop your files here, or",
        "description": "Text label inside the boundaries of the drag and drop area",
        "type": "string",
      },
      "fileNameLabel": {
        "default": "File",
        "description": "The header label for the column listing the individual file names",
        "type": "string",
      },
      "filePattern": {
        "default": "*",
        "description": "User provided valid file types as a list of comma delimited strings. Use '*' to accept all file types.",
        "examples": [
          "*",
          ".txt, .csv, .xlsx",
        ],
        "type": "string",
      },
      "imagePreviewSize": {
        "description": "The maximum width in pixels of image previews to show if previews are enabled",
        "type": "number",
      },
      "maxFileSizeMessage": {
        "description": "A quick tip explaining the file size limits of your File component",
        "type": "string",
      },
      "maxFileSizeWarning": {
        "description": "Error message below uploaded files that exceed the maximum size",
        "type": "string",
      },
      "maxNumberOfFiles": {
        "description": "Maximum accepted number of files",
        "type": "number",
      },
      "multiTabExcel": {
        "default": false,
        "description": "If user uploads a multi-tab Excel file, parse all tabs into one JSON structure",
        "type": "boolean",
      },
      "multiple": {
        "description": "Whether a user can select multiple files to upload",
        "type": "boolean",
      },
      "parseOnUpload": {
        "default": false,
        "description": "File data encoded as Base64 string as part of submission data",
        "type": "boolean",
      },
      "removeLink": {
        "default": "remove file",
        "description": "Aria-label passed to remove button inside uploaded files table",
        "type": "string",
      },
      "showImagePreview": {
        "description": "Whether uploaded image files should render a preview",
        "type": "boolean",
      },
      "sizeLabel": {
        "default": "Size",
        "description": "The header label for the column listing the file sizes",
        "type": "string",
      },
      "startingUploadMessage": {
        "default": "Starting upload",
        "description": "Text shown to users when upload begins",
        "type": "string",
      },
      "uqMaxFileSizeMB": {
        "default": "",
        "description": "Maximum file size in megabytes",
        "type": "object",
      },
    },
    "type": "object",
  },
  "FileStyleTargets": {
    "properties": {
      "acceptedFileList": {
        "description": "CSS target for the list of accepted files",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "dropRow": {
        "description": "CSS target for file drop row",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "dropTarget": {
        "description": "CSS target for file drop zone area",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "errors": {
        "description": "CSS target for field errors",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "fileBrowseButton": {
        "description": "CSS target for browse files button",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "fileError": {
        "description": "CSS target for errors on rejected files",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "imagePreview": {
        "description": "CSS target for file image previews",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "innerWrapper": {
        "description": "CSS target for file wrapper first child element",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "label": {
        "description": "CSS target for the label element",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "maxFileSize": {
        "description": "CSS target for the max file size text",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "maxSizeHelperText": {
        "description": "CSS target for file max size helper text",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "rejectedFileList": {
        "description": "CSS target for the list of rejected files",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "removeButton": {
        "description": "CSS target for file remove button",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "root": {
        "description": "CSS target for the file root element",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "tooltipButton": {
        "description": "CSS target for file tooltip button",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "tooltipPopover": {
        "description": "CSS target for file tooltip text",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "uploadIcon": {
        "description": "CSS target for file upload icon",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
    },
    "type": "object",
  },
  "FileStyling": {
    "description": "Holds information for targeted styling of the component.",
    "properties": {
      "targets": {
        "$ref": "#/definitions/FileStyleTargets",
        "type": "object",
      },
    },
    "type": "object",
  },
  "InputField": {
    "description": "Details to be applied to the input element within a component.",
    "properties": {
      "customClass": {
        "description": "Adds a custom CSS class to the component. This is useful for targeting this component to apply custom styles.",
        "type": "string",
      },
      "helperText": {
        "$ref": "#/definitions/InputHelperText",
      },
      "label": {
        "$ref": "#/definitions/InputLabel",
      },
      "placeholder": {
        "description": "An example or short explanation of the requested input. This appears inside of the input by default and disappears when the user enters a value.",
        "examples": [
          "First name - Expects the user to type their first name",
          "Enter your Social Security Number - Expects user's SSN",
          "dd/MM/YYYY - To show an expected date format (if you need a more robust component, try Date Input",
        ],
        "type": "string",
      },
      "prefix": {
        "description": "A symbol or text to display before the input field. Note this value will not be part of the submitted data",
        "examples": [
          "$",
          "@",
        ],
        "type": "string",
      },
      "suffix": {
        "description": "A symbol or text to display after the input field. Note this value will not be part of the submitted data",
        "examples": [
          ".com",
        ],
        "type": "string",
      },
      "tooltipDescription": {
        "description": "Tooltips are floating texts that appear when the cursor is positioned over an icon. They can span across multiple lines and provide hints about the requested data. Setting this property will make a question mark icon to be displayed on the right side of the component's label, and once the user hovers over it, the tooltip text will appear. Moving the cursor outside the icon's area hides the tooltip. Hiding the component's label automatically hides the tooltip.",
        "type": "string",
      },
    },
    "type": "object",
  },
  "InputHelperText": {
    "description": "Details pertaining to the helper text element on a component, including the text to be displayed and visibility.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "When \`true\`, it hides the helper text of the component.",
        "type": "boolean",
      },
      "text": {
        "description": "Configurable text that is displayed underneath the input field. Unlike tooltips, a helper text stays visible under the input field and displays on a single line.",
        "type": "string",
      },
    },
    "type": "object",
  },
  "InputLabel": {
    "description": "Details pertaining to the label element on a component, including the text to be displayed and visibility.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "When \`true\`, it hides the label of the component. Note the label will still be displayed on the Module Editor.",
        "type": "boolean",
      },
      "text": {
        "description": "Configurable text that is displayed above the input field.",
        "type": "string",
      },
    },
    "type": "object",
  },
  "InputRef": {
    "description": "Objects may define other targets that they refer to. This becomes a link for operation execution within the Runtime.",
    "properties": {
      "alias": {
        "type": "string",
      },
      "propertyPath": {
        "type": "string",
      },
      "required": {
        "type": "boolean",
      },
      "silent": {
        "type": "boolean",
      },
      "targetKey": {
        "minLength": 1,
        "type": "string",
      },
      "type": {
        "type": "string",
      },
    },
    "required": [
      "targetKey",
    ],
    "type": "object",
  },
  "Label": {
    "properties": {
      "text": {
        "description": "Text displayed as the label for the field",
        "type": "string",
      },
    },
    "type": "object",
  },
  "PackageModel": {
    "description": "The PackageModel describes how the Vega Runtime will resolve dependencies",
    "properties": {
      "type": {
        "description": "Type of package",
        "type": "string",
      },
      "url": {
        "description": "URL of package",
        "type": "string",
      },
      "version": {
        "description": "Version of package",
        "type": "string",
      },
    },
    "type": "object",
  },
  "StylingModel": {
    "description": "Holds information for targeted styling of the component.",
    "properties": {
      "targets": {
        "type": "object",
      },
    },
    "type": "object",
  },
  "Validation": {
    "description": "Holds the validation rules and the most current validation results for the component",
    "properties": {
      "isDirty": {
        "description": "Whether or not field has had user interaction. It will be set to true on user interaction.",
        "type": "boolean",
      },
      "result": {
        "$ref": "#/definitions/ValidationResult",
      },
      "rules": {
        "description": "The set of rules to restrict or define the accepted data for the current component. These are the rules to run for validation of the data input.",
        "examples": [
          "You can use it to define the max length of characters in an input field with a custom error message. rules: { required: { type: ValidationRuleTypes.MAX_LENGTH, validation: { maxLength: 5, }, errorMessage: 'Max length is 5', }, },",
        ],
        "type": "object",
      },
    },
    "type": "object",
  },
  "ValidationError": {
    "properties": {
      "message": {
        "description": "The error message",
        "type": "string",
      },
      "type": {
        "description": "Type of the error",
        "enum": [
          "required",
          "maxLength",
          "minLength",
          "arrayMinLength",
          "arrayMaxLength",
          "pattern",
          "min",
          "max",
          "mask",
          "number",
          "dateinput",
          "isBeforeDate",
          "isAfterDate",
          "customError",
          "isEmail",
        ],
        "minLength": 1,
        "type": "string",
      },
    },
    "required": [
      "type",
    ],
    "type": "object",
  },
  "ValidationResult": {
    "properties": {
      "errors": {
        "items": {
          "$ref": "#/definitions/ValidationError",
        },
        "type": "array",
      },
      "isValid": {
        "type": [
          "null",
          "boolean",
        ],
      },
    },
    "type": "object",
  },
}
`;
